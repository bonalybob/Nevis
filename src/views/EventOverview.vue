<template>
  <main>
    <h1 class="mx-10 mb-1">
      <back-arrow to="/events" />
      {{ event.name }}
      <template v-if="event.date">on {{ event.date }}</template>
    </h1>
    <div class="mx-12 mb-3">
      <router-link tag="button" class="button" :to="`/events/edit/${event.id}`">Edit Event</router-link>
    </div>
    <div class="mx-12 flex">
      <router-link
        tag="div"
        :to="`/events/${event.id}/courses`"
        class="shadow w-full md:w-6/12 text-center p-2 md:mr-4"
      >
        <h2>Courses</h2>
      </router-link>
    </div>
  </main>
</template>

<script>
import BackArrow from '@/components/BackArrow'

export default {
  components: {
    'back-arrow': BackArrow,
  },

  data: function () {
    return {
      event: {},
    }
  },

  mounted: function () {
    if (this.$database.connection === null || !this.$database.connected) this.$router.push('/')
    else this.getEventDetails()
  },

  methods: {
    getEventDetails: function () {
      return this.$database.query('SELECT * FROM events WHERE id=? LIMIT 1', this.$route.params.id)
        .then(result => { this.event = result[0] })
        .catch(error => this.$messages.addMessage(error, 'error'))
    },
  },
}
</script>

