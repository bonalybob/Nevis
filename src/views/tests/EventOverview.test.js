import { mount, shallowMount } from '@vue/test-utils'
import EventOverview from '@/views/EventOverview'

test('Is a Vue Instance', () => {
  const wrapper = shallowMount(EventOverview, {
    stubs: ['router-link'],
    mocks: {
      $database: { connection: {}, connected: true, query: jest.fn().mockResolvedValue() },
      $route: { params: { id: 12 } },
      $messages: { addMessage: jest.fn() },
    },
  })
  expect(wrapper.isVueInstance()).toBeTruthy()
})

test('Renders Correctly', () => {
  const wrapper = shallowMount(EventOverview, {
    stubs: ['router-link'],
    mocks: {
      $database: { connection: {}, connected: true, query: jest.fn().mockResolvedValue() },
      $route: { params: { id: 12 } },
      $messages: { addMessage: jest.fn() },
    },
  })
  wrapper.setData({ event: { id: 12, name: 'Test Event', date: '11/22/33' } })
  expect(wrapper.element).toMatchSnapshot()
})

test('Not Connected to the Database', () => {
  const wrapper = shallowMount(EventOverview, {
    stubs: ['router-link'],
    mocks: {
      $database: { connection: {}, connected: false, query: jest.fn().mockResolvedValue() },
      $route: { params: { id: 12 } },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn() },
    },
  })
  expect(wrapper.vm.$router.push).toHaveBeenCalledTimes(1)
  expect(wrapper.vm.$router.push).toHaveBeenLastCalledWith('/')
})

test('Get Event Details - Success', async () => {
  const wrapper = mount(EventOverview, {
    stubs: ['router-link'],
    mocks: {
      $database: { connection: {}, connected: true, query: jest.fn().mockResolvedValue(['hello']) },
      $route: { params: { id: 12 } },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn() },
    },
  })
  await wrapper.vm.getEventDetails()
  expect(wrapper.vm.event).toBe('hello')
})

test('Get Event Details - Error', async () => {
  const wrapper = mount(EventOverview, {
    stubs: ['router-link'],
    mocks: {
      $database: { connection: {}, connected: true, query: jest.fn().mockRejectedValue() },
      $route: { params: { id: 12 } },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn() },
    },
  })
  await wrapper.vm.getEventDetails()
  expect(wrapper.vm.$messages.addMessage).toHaveBeenLastCalledWith('Problem Fetching Event Data', 'error')
  expect(wrapper.vm.event).toEqual({})
})
