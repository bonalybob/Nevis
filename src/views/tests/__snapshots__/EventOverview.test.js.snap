// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders Correctly 1`] = `
<main>
  <h1
    class="mx-10 mb-1"
  >
    <back-arrow-stub
      to="/events"
    />
    
    Test Event
    on 11/22/33
  </h1>
   
  <div
    class="mx-12 mb-3"
  >
    <router-link-stub
      class="button"
      tag="button"
      to="/events/edit/12"
    >
      Edit Event
    </router-link-stub>
  </div>
   
  <div
    class="mx-12 flex"
  >
    <router-link-stub
      class="shadow w-full md:w-6/12 text-center p-2 md:mr-4"
      tag="div"
      to="/events/12/courses"
    >
      <h2>
        Courses
      </h2>
    </router-link-stub>
  </div>
</main>
`;
